README.md
[![NPM](https://nodei.co/npm/yocto-status-codes.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/yocto-status-codes/)

![alt text](https://david-dm.org/yoctore/yocto-status-codes.svg "Dependencies Status")
[![Code Climate](https://codeclimate.com/github/yoctore/yocto-status-codes/badges/gpa.svg)](https://codeclimate.com/github/yoctore/yocto-status-codes)
[![Test Coverage](https://codeclimate.com/github/yoctore/yocto-status-codes/badges/coverage.svg)](https://codeclimate.com/github/yoctore/yocto-status-codes/coverage)
[![Issue Count](https://codeclimate.com/github/yoctore/yocto-status-codes/badges/issue_count.svg)](https://codeclimate.com/github/yoctore/yocto-status-codes)
[![Build Status](https://travis-ci.org/yoctore/yocto-status-codes.svg?branch=master)](https://travis-ci.org/yoctore/yocto-status-codes)

## Overview

This module is a part of yocto node modules for NodeJS.

Please see [our NPM repository](https://www.npmjs.com/~yocto) for complete list of available tools (completed day after day).

Manage and retrieve status code & more defined data from given rules.

## Motivation

Create a standard response format for our [yocto-core-stack](https://www.npmjs.com/package/yocto-core-stack) application

## Which format have response ?

All reponse given by this module has format below :

```javascript
{
  status  : "success|error", // Always success or error
  code    : "200|400|500", // At the moment only this three values
  message : "", // a specific message, can be overload
  data    : {} // a specific data, can be overload
}
```

## Configuration file

The configuration file give you the possibility to load predefined data in an jSON file.

Childs categories can only up to 999 like 200999

### Structure of file :

```javascript
[
  // define some parents categories
  {
    "category"  : 200,
    // define some childs categories
    "childs"    :  [
      {
        "code"      : "000",
        "codeHTTP"  : 200,
        "status"    : "success",
        "label"     : "success-base-code",
        "message"   : "Success base code"
      }
    ]
  },
  {
    "category"  : 206,
    // Define childs for level 2060xx like 206000
    "childs"    :  [
      {
        "code"      : "000",
        "codeHTTP"  : 200,
        "status"    : "success",
        "label"     : "success-partial",
        "message"   : "default"
      },
      {
        "category" : 100,
        // Define childs for level 2061xx like 206101
        "childs"   :  [
          {
            "code"      : "101",
            "status"    : "error",
            "codeHTTP"  : 200,
            "label"     : "account-base-error-code",
            "message"   : ""
          }
        ]
      }
    ]
  }
]
```

## How to use

At the moment we only have 5 method implemented, see below :

```javascript

var status = require('yocto-status-codes');

var successData = { message : 'Yeah !', data : { foo : 'bar' } };
var errorData = { message : 'Bad !', data : { foo : 'bar error' } };
var systemErrorData = { message : 'Very bad !', data : { foo : 'bar system' } };

// To retrieve a success reponse
status.success(successData);

// To add a specific message on base response just do
status.success(successData, 'message');
// it works too for error and system erro

// To retrieve an error reponse
status.error(errorData);

// To retrieve an system error reponse
status.systemError(systemErrorData);

// Load an JSON config file by passing his path
status.loadConfig('/home/config.json');

// Get an existing StatusCode from configuration file
status.get(200, 100);

// Get an existing StatusCode and override field "data" or/and "message"
status.get(200, 100, {
  message : 'overriden message', // optional field
  data    : [ 'this data was overriden' ] // optional field
});
```


## Changelog

All history is [here](https://github.com/yoctore/yocto-status-codes/blob/master/CHANGELOG.md)



CHANGELOG.md
## 1.2.0 && 1.2.1 && 1.2.2 (2016-03-30)
- Add function to load an configuration file and get status code by reference code
- Add test units
- Add joi validation when loading configuration file
- Add possibilities to set yocto-logger in constructor
- Disale logger when testing

## 1.1.0 && 1.1.1 && 1.1.2 (2015-12-21)

- Add a param to overload message value
- Add default value on properties if is not set. in some case data is not clean

## 1.0.1

- Fix a merge for data

## 1.0.0

- Base implementation & tests
- Add success / error / system error generator methods


test/test.js
var logger      = require('yocto-logger');
var statusCodes = require('../src/index')(logger);
var path        = require('path');
var base        = path.normalize(process.cwd());
var chai        = require('chai');
var assert      = chai.assert;

// disableConsole
logger.disableConsole();

describe('Test module', function() {

  /**
  * try to retreive an status codes without Initialising configuration
  */
  it('should return a default header because config was not initalised', function () {

    var res = statusCodes.get(206);

    assert.equal(res.codeHTTP, 200);
    assert.equal(res.content.code, '404404');
    assert.equal(res.content.status, 'error');
    assert.equal(res.content.message, 'Cannot retrieve current header');
  });

  /**
  * try to load an wrong file
  */
  it('should return false when loading config because file doesnt exist', function () {

    // load configuration
    var res =   statusCodes.loadConfig(base + '/test/configNotExist.json');

    assert.equal(res, false);
  });

  /**
  * try to load an wrong file
  */
  it('should return false when loading config because file doesnt have a good schema', function () {

    // load configuration
    var res =   statusCodes.loadConfig(base + '/test/configError.json');

    assert.equal(res, false);
  });

  /**
  * try to load config file
  */
  it('should return true when loading file ', function () {

    // load configuration
    var res =   statusCodes.loadConfig(base + '/test/config.json');

    assert.equal(res, true);
  });

  /**
  * try retrieve code 206000
  */
  it('should return code 206201 in config file', function () {

    // retrieve code
    var res =   statusCodes.get(206, 201);

    assert.equal(res.codeHTTP, 200);
    assert.equal(res.content.code, '206201');
    assert.equal(res.content.status, 'error');
    assert.equal(res.content.message, 'test code 201');
  });

  /**
  * try retrieve code 206000
  */
  it('should return code 206000 found in config file', function () {

    // retrieve code
    var res =   statusCodes.get(206);

    assert.equal(res.codeHTTP, 200);
    assert.equal(res.content.code, '206000');
    assert.equal(res.content.status, 'success');
    assert.equal(res.content.message, 'default');
  });

  /**
  * try to retreive an status codes without Initialising configuration
  */
  it('should return a default header because code specified doesnt exist ', function () {

    // retrieve code
    var res = statusCodes.get(209);

    assert.equal(res.codeHTTP, 200);
    assert.equal(res.content.code, '404404');
    assert.equal(res.content.status, 'error');
    assert.equal(res.content.message, 'Cannot retrieve current header');
  });

  /**
  * try retrieve code 206000 with overiden message
  */
  it('should return code 206 with overriden message', function () {

    // retrieve code
    var res =   statusCodes.get(206, 0, { message : 'overriden' });

    assert.equal(res.codeHTTP, 200);
    assert.equal(res.content.code, '206000');
    assert.equal(res.content.status, 'success');
    assert.equal(res.content.message, 'overriden');
  });

  /**
  * try retrieve code 206000 with overiden data
  */
  it('should return code 206 with overriden data', function () {

    // retrieve code
    var res =   statusCodes.get(206, 0, {
      data : {
        overriden : true
      }
    });

    assert.equal(res.codeHTTP, 200);
    assert.equal(res.content.code, '206000');
    assert.equal(res.content.status, 'success');
    assert.equal(res.content.data.overriden, true);
  });

  /**
  * try retrieve code 206000 with overiden data and message
  */
  it('should return code 206 with overriden data and message', function () {

    // retrieve code
    var res =   statusCodes.get(206, 0, {
      data      : {
        overriden : true
      },
      message   : 'overriden'
    });

    assert.equal(res.codeHTTP, 200);
    assert.equal(res.content.code, '206000');
    assert.equal(res.content.status, 'success');
    assert.equal(res.content.data.overriden, true);
    assert.equal(res.content.message, 'overriden');
  });

  /**
  * try to retreive an status codes with wrong data
  */
  it('should return a default header because data is wrong ', function () {

    // retrieve code
    var res = statusCodes.get(206,0, {befe : 'fe'});

    assert.equal(res.codeHTTP, 200);
    assert.equal(res.content.code, '404404');
    assert.equal(res.content.status, 'error');
    assert.equal(res.content.message, 'Cannot retrieve current header');
  });
});


Gruntfile.js
'use strict';

module.exports = function (grunt) {
  // init config
  grunt.initConfig({
    // default package
    pkg       : grunt.file.readJSON('package.json'),
    // hint our app
    yoctohint : {
      options  : {},
      all      : [ 'src/***', 'Gruntfile.js' ]
    },
    // Uglify our app
    uglify    : {
      options : {
        banner  : '/* <%= pkg.name %> - <%= pkg.description %> - V<%= pkg.version %> */\n'
      },
      api     : {
        files : [ {
          expand  : true,
          cwd     : 'src',
          src     : '**/*.js',
          dest    : 'dist'
        } ]
      }
    },
    // unit testing
    mochaTest : {
      // Test all unit test
      all  : {
        options : {
          reporter : 'list',
        },
        src     : [ 'test/*.js' ]
      }
    }
  });

  // Load the plugins
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-mocha-test');
  grunt.loadNpmTasks('yocto-hint');

  grunt.registerTask('hint', 'yoctohint');
  grunt.registerTask('test', 'mochaTest');
  grunt.registerTask('build', [ 'hint', 'test', 'uglify' ]);
  grunt.registerTask('default', 'build');
};


src/index.js
'use strict';

var _       = require('lodash');
var logger  = require('yocto-logger');
var path    = require('path');
var fs      = require('fs');
var joi     = require('joi');

/**
 * Manage status code response
 *
 * @class Status Codes
 */
function StatusCodes (yLogger) {
  /**
   * Default status rules
   */
  this.defaultStatus = {
    // success item
    success : {
      status : 'success',
      code   : '200'
    },
    // error item
    error   : {
      status : 'error',
      code   : '400'
    },
    // system error item
    system  : {
      status : 'error',
      code   : '500'
    }
  };

  /**
   * Contain external file for error
   */
  this.file = {};

  /**
   * Default logger
   */
  this.logger = yLogger || logger;
}

/**
 * Default method to build default object from given rules
 *
 * @param {String} type default wanted type
 * @param {Object} data default data to append on current object
 * @param {String} message specific message to ad on default object
 * @return {Object|Boolean} builded object false if is invalid
 */
StatusCodes.prototype.buildDefault = function (type, data, message) {
  // type exists ?
  if (_.has(this.defaultStatus, type) && _.isString(type)) {
    // default statement
    var d  = _.extend({}, _.clone(this.defaultStatus[type]), { message : '', data : {} });

    // given data is object ?
    if (_.isObject(data)) {
      // extend data with correct values
      _.extend(d.data, data);
    }

    // assign specific message
    if (_.isString(message) && !_.isEmpty(message)) {
      // change message
      d.message = message;
    }

    // default statement
    return d;
  }

  // invalid statement
  return false;
};

/**
 * Default method to build a valid response object object
 *
 * @param {Object} data default data to append on current object
 * @param {String} message specific message to ad on default object
 * @return {Object|false} builded object false if is invalid
 */
StatusCodes.prototype.success = function (data, message) {
  // default statement
  return this.buildDefault('success', data || {}, message || '');
};

/**
 * Default method to build an error response object object
 *
 * @param {Object} data default data to append on current object
 * @param {String} message specific message to ad on default object
 * @return {Object|false} builded object false if is invalid
 */
StatusCodes.prototype.error = function (data, message) {
  // default statement
  return this.buildDefault('error', data || {}, message || '');
};

/**
 * Default method to build an system response error object object
 *
 * @param {Object} data default data to append on current object
 * @param {String} message specific message to ad on default object
 * @return {Object|false} builded object false if is invalid
 */
StatusCodes.prototype.systemError = function (data, message) {
  // default statement
  return this.buildDefault('system', data || {}, message || '');
};

/**
 * Load an external file to construct response just by specifing error code
 *
 * @param  {String} pathFile the path of the config file
 * @return {boolean} true if file is loaded, otherwise false
 */
StatusCodes.prototype.loadConfig = function (pathFile) {
  // try to load external file
  try {

    // Parse the file
    this.file = JSON.parse(fs.readFileSync(path.normalize(pathFile)));

    // joi schema of the config file
    var schema = joi.array().min(1).items(
      joi.object().required().keys({
        category    : joi.number().integer().required().min(0),
        childs      :  joi.array().required().items(
          joi.alternatives().try(
            joi.object().keys({
              code      : joi.number().integer().required(),
              codeHTTP  : joi.number().integer().required(),
              status    : joi.string().required().empty(),
              label     : joi.string().required().empty(),
              message   : joi.string().required().empty()
            }),
            joi.object().keys({
              category    : joi.number().integer().required().min(0),
              childs      :  joi.array().required().items(
                joi.object().keys({
                  code      : joi.number().integer().required(),
                  codeHTTP  : joi.number().integer().required(),
                  status    : joi.string().required().empty(),
                  label     : joi.string().required().empty(),
                  message   : joi.string().required().empty()
                })
              )
            })
          )
        )
      })
    );

    // validate joi schema with the given file
    var result   = schema.validate(this.file);

    // check if an error occured
    if (result.error) {
      // throw a new exception
      throw ('The joi validation failed, more details : ' + result.error.toString());
    }

    this.logger.info('[ StatusCodes.loadConfig ] - an external configuration file : ' + pathFile +
    ' was loaded with success');
    // Config file was success loaded
    return true;
  } catch (error) {
    // An error occured when loading file
    this.logger.error('[ StatusCodes.loadConfig ] - an error occured when loading external ' +
    'configuration file, more details : ' + error);
    // Config file was not loaded
    return false;
  }
};

/**
 * Retrieve status object from loaded config file
 *
 * @param  {Integer} category  The main category of code (ex : 200 or 400)
 * @param  {String} code       The sub category of code (ex : '101')
 * @return {Object}            Return an object with data of code
 */
StatusCodes.prototype.get = function (category, code, data) {

  // Initialise subCode if is not define
  code = code || '000';

  // Set An default error code to indicate that an error occured and response can't be seet
  var errorResponse = {
    codeHTTP    : 200,
    content     : {
      code      : '404404',
      status    : 'error',
      message   : 'Cannot retrieve current header',
      data      : {}
    }
  };

  // check if an config file exist
  if (_.isEmpty(this.file)) {

    // log error
    this.logger.error('[ response.get ] - error when retriving response, more details :' +
    ' configuration file was not loaded');

    // return error message
    return errorResponse;
  }

  // Schema to check given params
  var schema =  joi.object().keys({
    category : joi.number().required().empty().min(0),
    code     : joi.number().required().empty().min(0).max(999),
    data     : joi.object().optional().keys({
      data    : joi.any().optional().default({}),
      message : joi.string().optional()
    }).default({ data : {} })
  });

  // test if schema of object was ok
  var result = joi.validate({
    category    : category,
    code        : code,
    data        : data
  }, schema);

  // parse the Integer code into string that have 3 character
  code = ('00' + code).slice(-3);

  // override data with default value from joi schema
  data = result.value.data;

  // define final object to return
  var response = {};

  // try to create final object
  try {

    // Test if joi validation failed
    if (!_.isEmpty(result.error)) {
      throw result.error;
    }

    // Retrieve and clone the main category in file
    var mainCategoryFound     = _.clone(_.find(this.file, { 'category' : category })) || [];

    // Test if is main category
    if (_.parseInt(code) <= 100) {
      response  = _.clone(_.find(mainCategoryFound.childs, { code : code }));
    } else {

      // found subcategory
      var subcategoryFound  = _.clone(_.find(mainCategoryFound.childs,
        { category : _.floor(_.parseInt(code), -2) }) || []
      );

      // clone response
      response = _.clone(_.find(subcategoryFound.childs, { code : code }) || {});
    }

    // If undefined return the default header error
    if (_.isUndefined(response) || _.isEmpty(response)) {
      // return an default error message
      return errorResponse;
    }

    // Set code error
    response.code = category.toString() + response.code;

    // return the final object
    response = {
      codeHTTP : response.codeHTTP,
      // omit uncessary keys form config
      content  : _.omit(response, [ 'label', 'codeHTTP' ])
    };

    // set content key
    response.content = _.extend(response.content, data);

    // return the response
    return response;
  } catch (error) {
    // set error response
    response = errorResponse;

    // An error occured
    this.logger.error('[ response.get ] - error when retriving response, more details :', error);

    // return the response
    return response;
  }
};

// Default export
module.exports = function (l) {
  // is a valid logger ?
  if (_.isUndefined(l) || _.isNull(l)) {
    logger.warning('[ Yocto-status-code.constructor ] - Invalid logger given. Use internal logger');
    // assign
    l = logger;
  }
  // default statement
  return new (StatusCodes)(l);
};
